File: expo-refrence/src/crypto/nip44.ts
import { chacha20 } from '@noble/ciphers/chacha.js';
import { expand, extract } from '@noble/hashes/hkdf.js';
import { hmac } from '@noble/hashes/hmac.js';
import { sha256 } from '@noble/hashes/sha2.js';
import { Point, utils } from '@noble/secp256k1';
import { base64 } from '@scure/base';

import { bytesToUtf8, concatBytes, hexToBytes, utf8ToBytes } from '../utils/encoding';

const VERSION = 2;
const MIN_PLAINTEXT_SIZE = 1;
const MAX_PLAINTEXT_SIZE = 65535;
const CONVERSATION_SALT = utf8ToBytes('nip44-v2');

function bytesToNumber(bytes: Uint8Array): bigint {
  let value = 0n;

  for (const byte of bytes) {
    value = (value << 8n) | BigInt(byte);
  }

  return value;
}

function numberToBytes(value: number): Uint8Array {
  return Uint8Array.of((value >> 8) & 0xff, value & 0xff);
}

function equalBytes(left: Uint8Array, right: Uint8Array): boolean {
  if (left.length !== right.length) {
    return false;
  }

  let result = 0;

  for (let index = 0; index < left.length; index += 1) {
    result |= left[index] ^ right[index];
  }

  return result === 0;
}

function calcPaddedLength(unpaddedLength: number): number {
  if (unpaddedLength <= 32) {
    return 32;
  }

  const nextPower = 1 << (Math.floor(Math.log2(unpaddedLength - 1)) + 1);
  const chunk = nextPower <= 256 ? 32 : nextPower / 8;

  return chunk * (Math.floor((unpaddedLength - 1) / chunk) + 1);
}

function pad(plaintext: string): Uint8Array {
  const unpadded = utf8ToBytes(plaintext);
  const unpaddedLength = unpadded.length;

  if (unpaddedLength < MIN_PLAINTEXT_SIZE || unpaddedLength > MAX_PLAINTEXT_SIZE) {
    throw new Error('Invalid plaintext length');
  }

  return concatBytes(numberToBytes(unpaddedLength), unpadded, new Uint8Array(calcPaddedLength(unpaddedLength) - unpaddedLength));
}

function unpad(padded: Uint8Array): string {
  if (padded.length < 34) {
    throw new Error('Invalid padded payload');
  }

  const unpaddedLength = (padded[0] << 8) | padded[1];
  const plaintext = padded.slice(2, 2 + unpaddedLength);

  if (
    unpaddedLength < MIN_PLAINTEXT_SIZE ||
    plaintext.length !== unpaddedLength ||
    padded.length !== 2 + calcPaddedLength(unpaddedLength)
  ) {
    throw new Error('Invalid padding');
  }

  return bytesToUtf8(plaintext);
}

function getConversationKey(privkey: string, pubkey: string): Uint8Array {
  return extract(sha256, computeSharedSecret(privkey, pubkey), CONVERSATION_SALT);
}

function getMessageKeys(conversationKey: Uint8Array, nonce: Uint8Array) {
  if (conversationKey.length !== 32 || nonce.length !== 32) {
    throw new Error('Invalid key material');
  }

  const keys = expand(sha256, conversationKey, nonce, 76);

  return {
    chachaKey: keys.slice(0, 32),
    chachaNonce: keys.slice(32, 44),
    hmacKey: keys.slice(44, 76),
  };
}

function decodePayload(payload: string) {
  if (payload.length === 0 || payload.startsWith('#')) {
    throw new Error('Unsupported payload version');
  }

  if (payload.length < 132 || payload.length > 87472) {
    throw new Error('Invalid payload size');
  }

  const data = base64.decode(payload);

  if (data.length < 99 || data.length > 65603) {
    throw new Error('Invalid decoded payload size');
  }

  if (data[0] !== VERSION) {
    throw new Error('Unsupported payload version');
  }

  return {
    nonce: data.slice(1, 33),
    ciphertext: data.slice(33, data.length - 32),
    mac: data.slice(data.length - 32),
  };
}

export function computeSharedSecret(privkey: string, pubkey: string): Uint8Array {
  const privateKeyBytes = hexToBytes(privkey);

  if (!utils.isValidSecretKey(privateKeyBytes)) {
    throw new Error('Invalid private key');
  }

  const point = Point.fromHex(`02${pubkey}`);

  return point.multiply(bytesToNumber(privateKeyBytes)).toBytes(true).slice(1);
}

export function nip44Encrypt(
  plaintext: string,
  senderPrivkey: string,
  recipientPubkey: string,
): string {
  const conversationKey = getConversationKey(senderPrivkey, recipientPubkey);
  const nonce = utils.randomSecretKey();
  const { chachaKey, chachaNonce, hmacKey } = getMessageKeys(conversationKey, nonce);
  const ciphertext = chacha20(chachaKey, chachaNonce, pad(plaintext));
  const mac = hmac(sha256, hmacKey, concatBytes(nonce, ciphertext));

  return base64.encode(concatBytes(Uint8Array.of(VERSION), nonce, ciphertext, mac));
}

export function nip44Decrypt(
  ciphertext: string,
  recipientPrivkey: string,
  senderPubkey: string,
): string {
  const { nonce, ciphertext: encrypted, mac } = decodePayload(ciphertext);
  const conversationKey = getConversationKey(recipientPrivkey, senderPubkey);
  const { chachaKey, chachaNonce, hmacKey } = getMessageKeys(conversationKey, nonce);
  const calculatedMac = hmac(sha256, hmacKey, concatBytes(nonce, encrypted));

  if (!equalBytes(calculatedMac, mac)) {
    throw new Error('Invalid MAC');
  }

  return unpad(chacha20(chachaKey, chachaNonce, encrypted));
}

File: app/src/main/java/com/ely/kian/crypto/Nip44.kt
package com.ely.kian.crypto

import android.util.Base64
import java.nio.ByteBuffer
import java.nio.ByteOrder
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlin.math.ln
import kotlin.math.pow

object Nip44 {
    private const val VERSION = 2
    private val CONVERSATION_SALT = "nip44-v2".toByteArray()

    fun encrypt(plaintext: String, senderPrivKey: ByteArray, recipientPubKey: ByteArray): String {
        val sharedSecret = KianKeys.computeSharedSecret(senderPrivKey, recipientPubKey)
        val conversationKey = hkdfExtract(sharedSecret, CONVERSATION_SALT)
        
        val nonce = ByteArray(32)
        java.security.SecureRandom().nextBytes(nonce)
        
        val keys = getMessageKeys(conversationKey, nonce)
        val padded = pad(plaintext)
        
        val cipher = Cipher.getInstance("ChaCha20")
        val keySpec = SecretKeySpec(keys.chachaKey, "ChaCha20")
        // NIP-44 uses 12-byte nonce for ChaCha20
        val ivSpec = IvParameterSpec(keys.chachaNonce)
        cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec)
        
        val ciphertext = cipher.doFinal(padded)
        val hmac = Mac.getInstance("HmacSHA256")
        hmac.init(SecretKeySpec(keys.hmacKey, "HmacSHA256"))
        hmac.update(nonce)
        val mac = hmac.doFinal(ciphertext)
        
        val result = ByteBuffer.allocate(1 + nonce.size + ciphertext.size + mac.size)
        result.put(VERSION.toByte())
        result.put(nonce)
        result.put(ciphertext)
        result.put(mac)
        
        return Base64.encodeToString(result.array(), Base64.NO_WRAP)
    }

    fun decrypt(payload: String, recipientPrivKey: ByteArray, senderPubKey: ByteArray): String {
        val data = Base64.decode(payload, Base64.NO_WRAP)
        if (data[0].toInt() != VERSION) throw Exception("Unsupported NIP-44 version")
        
        val nonce = data.sliceArray(1..32)
        val mac = data.sliceArray(data.size - 32 until data.size)
        val ciphertext = data.sliceArray(33 until data.size - 32)
        
        val sharedSecret = KianKeys.computeSharedSecret(recipientPrivKey, senderPubKey)
        val conversationKey = hkdfExtract(sharedSecret, CONVERSATION_SALT)
        val keys = getMessageKeys(conversationKey, nonce)
        
        val hmac = Mac.getInstance("HmacSHA256")
        hmac.init(SecretKeySpec(keys.hmacKey, "HmacSHA256"))
        hmac.update(nonce)
        val calculatedMac = hmac.doFinal(ciphertext)
        
        if (!calculatedMac.contentEquals(mac)) throw Exception("Invalid NIP-44 MAC")
        
        val cipher = Cipher.getInstance("ChaCha20")
        val keySpec = SecretKeySpec(keys.chachaKey, "ChaCha20")
        val ivSpec = IvParameterSpec(keys.chachaNonce)
        cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec)
        
        val padded = cipher.doFinal(ciphertext)
        return unpad(padded)
    }

    private data class MessageKeys(val chachaKey: ByteArray, val chachaNonce: ByteArray, val hmacKey: ByteArray)

    private fun getMessageKeys(conversationKey: ByteArray, nonce: ByteArray): MessageKeys {
        val keys = hkdfExpand(conversationKey, nonce, 76)
        return MessageKeys(
            chachaKey = keys.sliceArray(0..31),
            chachaNonce = keys.sliceArray(32..43),
            hmacKey = keys.sliceArray(44..75)
        )
    }

    private fun hkdfExtract(ikm: ByteArray, salt: ByteArray): ByteArray {
        val hmac = Mac.getInstance("HmacSHA256")
        hmac.init(SecretKeySpec(salt, "HmacSHA256"))
        return hmac.doFinal(ikm)
    }

    private fun hkdfExpand(prk: ByteArray, info: ByteArray, length: Int): ByteArray {
        val hmac = Mac.getInstance("HmacSHA256")
        hmac.init(SecretKeySpec(prk, "HmacSHA256"))
        val okm = ByteArray(length)
        var t = ByteArray(0)
        var i = 1
        var pos = 0
        while (pos < length) {
            hmac.update(t)
            hmac.update(info)
            hmac.update(i.toByte())
            t = hmac.doFinal()
            val copyLen = minOf(t.size, length - pos)
            System.arraycopy(t, 0, okm, pos, copyLen)
            pos += copyLen
            i++
        }
        return okm
    }

    private fun pad(plaintext: String): ByteArray {
        val unpadded = plaintext.toByteArray()
        val len = unpadded.size
        if (len < 1 || len > 65535) throw Exception("Invalid plaintext length")
        
        val paddedLen = calcPaddedLength(len)
        val result = ByteArray(2 + paddedLen)
        result[0] = (len shr 8).toByte()
        result[1] = (len and 0xFF).toByte()
        System.arraycopy(unpadded, 0, result, 2, len)
        return result
    }

    private fun unpad(padded: ByteArray): String {
        val len = ((padded[0].toInt() and 0xFF) shl 8) or (padded[1].toInt() and 0xFF)
        if (len < 1 || len > padded.size - 2) throw Exception("Invalid padding")
        val plaintext = ByteArray(len)
        System.arraycopy(padded, 2, plaintext, 0, len)
        return String(plaintext)
    }

    private fun calcPaddedLength(len: Int): Int {
        if (len <= 32) return 32
        val nextPower = 1 shl (kotlin.math.floor(kotlin.math.log2(len - 1f)) + 1).toInt()
        val chunk = if (nextPower <= 256) 32 else nextPower / 8
        return chunk * (kotlin.math.floor((len - 1f) / chunk).toInt() + 1)
    }
}
